### ISSUE TITLE
Defect: tests fail when compiling OpenCoarrays (2.10.2) with MPICH (4.1.2)

### ISSUE THREAD

**Issue by banana-bred** [sourceryinstitute/OpenCoarrays#783]
[ X ] I am reporting a bug others will be able to reproduce and not asking a question or requesting a new feature.

## System information including:
  - OpenCoarrays Version: 
```
OpenCoarrays Coarray Fortran Compiler Wrapper (caf version 2.10.2)
```

  - Fortran Compiler: 
  ```
Using built-in specs.
COLLECT_GCC=gfortran
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-pc-linux-gnu/13.2.1/lto-wrapper
Target: x86_64-pc-linux-gnu
Configured with: /build/gcc/src/gcc/configure --enable-languages=ada,c,c++,d,fortran,go,lto,objc,obj-c++ --enable-bootstrap --prefix=/usr --libdir=/usr/lib --libexecdir=/usr/lib --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=https://bugs.archlinux.org/ --with-build-config=bootstrap-lto --with-linker-hash-style=gnu --with-system-zlib --enable-__cxa_atexit --enable-cet=auto --enable-checking=release --enable-clocale=gnu --enable-default-pie --enable-default-ssp --enable-gnu-indirect-function --enable-gnu-unique-object --enable-libstdcxx-backtrace --enable-link-serialization=1 --enable-linker-build-id --enable-lto --enable-multilib --enable-plugin --enable-shared --enable-threads=posix --disable-libssp --disable-libstdcxx-pch --disable-werror
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 13.2.1 20230801 (GCC) 
```
  - C compiler used for building lib: 
 ```
gcc (GCC) 13.2.1 20230801
```

  - Installation method: Followed the `cmake` build instructions : 
```
  cd <PACKAGE>

  export FC="$(command -v gfortran)"
  export CC="$(command -v gcc)"
  cmake \
    -B build \
    -DCMAKE_INSTALL_PREFIX=/usr \
    -DCMAKE_INSTALL_LIBDIR=lib \
    -DMPI_C_COMPILER=/opt/mpich/bin/mpicc \
    -DMPI_Fortran_COMPILER=/opt/mpich/bin/mpifort \
    -DMPIEXEC_EXECUTABLE=/opt/mpich/bin/mpiexec

  cd build

  make
```
  - All flags & options passed to the installer : See above
  - Output of `uname -a`:
  `Linux Tchaikovsky 6.7.1-artix1-1 #1 SMP PREEMPT_DYNAMIC Mon, 22 Jan 2024 16:10:38 +0000 x86_64 GNU/Linux
`
  - MPI library being used: MPICH 4.1.2
  - Machine architecture and number of physical cores: x86_64
  ```
$ lscpu | grep -E "Thread|Core|Socket|^CPU\(s\):"

CPU(s):                             16
Thread(s) per core:                 2
Core(s) per socket:                 8
Socket(s):                          1
```

  - Version of CMake: 3.28.1

## To help us debug your issue please explain:

Tests fail when compiling OpenCoarrays with MPICH : 
  - sync_team
  - sendget_convert_nums
  - sendget_convert_char_array
  - issue-503-multidim-array-broadcast

### What you were trying to do (and why)

Compile OpenCoarrays with MPICH.

### What happened (include command output, screenshots, logs, etc.)

Tests failed. See attached files for test output. 

### What you expected to happen

Tests succeed.

### Step-by-step reproduction instructions to reproduce the error/bug

Install mpich
[compilation.txt](https://github.com/sourceryinstitute/OpenCoarrays/files/14213494/compilation.txt)
[LastTest.log](https://github.com/sourceryinstitute/OpenCoarrays/files/14213495/LastTest.log)
, compile with above steps.

[links]:#
[fenced code blocks]: https://help.github.com/articles/creating-and-highlighting-code-blocks/
[CONTRIBUTING.md]: https://github.com/sourceryinstitute/OpenCoarrays/blob/master/CONTRIBUTING.md

**Comment by rouson** [comment]
@banana-bred in case you're still interested, I just completed an installation using GCC 14.1.0 and MPICH 4.2.2. I'll do some more testing and updated the OpenCoarrays installation scripts to use these versions.

**Comment by rouson** [comment]
All tests pass when I execute something like the following
```
export FC=`which gfortran-14`
export CC=`which gcc-14`
export CXX=`which g++-14`

./install.sh \
  /-p mpich \ 
  -i /opt/mpich/4.2.2/gnu/14.1.1 \
  -f $FC -c $CC -C $CXX \
  -I 4.2.2

export PATH=/opt/mpich/4.2.2/gnu/14.1.1/bin:PATH
export LD_LIBRARY_PATH=/opt/mpich/4.2.2/gnu/14.1.1/lib

./install.sh \
  -i /opt/opencoarrays/2.10.2/gnu/14.1.0 
  -f $FC -c $CC -C $CXX \
  -M /opt/mpich/4.2.2/gnu/14.1.0

source /opt/opencoarrays/2.10.2/gnu/14.1.0/setup.sh
cd prerequisites/builds/opencoarrays/2.10.2
ctest
```

**Comment by mkowsiak** [comment]
Will give it a try. Thanks.

**Comment by vehre** [comment]
This looks like a duplicate of #781 . At least it should be resolved by the pr #789 fixing #781.
